### PR TITLE
Add endpoint to device registry for package stats

### PR DESCRIPTION
**Pull request by CalumMcCall**
- PRO-1953

### CODE REVIEW

File: device-registry/src/main/scala/org/genivi/sota/device_registry/DevicePackagesResource.scala [hunk 1]
```diff
@@ -49,6 +57,9 @@ class DevicePackagesResource(namespaceExtractor: Directive1[AuthedNamespaceScope
           }
         }
       }
+    } ~
+    (scope.get & pathPrefix("package") & extractPackageId & path("stats")) { pkg =>
```
Comment (jerrytrieu) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/DevicePackagesResource.scala]: Let's not have 'package' path, keep it about devices. Eg 'devices_count' with parameter 'withPackage' etc.

File: device-registry/src/main/scala/org/genivi/sota/device_registry/db/InstalledPackages.scala [hunk 2]
```diff
@@ -22,6 +22,8 @@ object InstalledPackages {
   case class InstalledPackage(device: Uuid, packageId: PackageId,
                               lastModified: Instant)
 
+  case class PackageStats(deviceCount: Int, groupCount: Int)
```
Comment (jerrytrieu) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/InstalledPackages.scala]: Better 'DevicesCount'

File: device-registry/src/main/scala/org/genivi/sota/device_registry/db/InstalledPackages.scala [hunk 3]
```diff
@@ -56,4 +58,22 @@ object InstalledPackages {
       .filter(_.device === device)
       .result
 
+  def getPackageStats(pkg: PackageId, ns: Namespace)(implicit ec: ExecutionContext): DBIO[PackageStats] =
```
Comment (jerrytrieu) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/InstalledPackages.scala]: Same

File: common-client/src/main/scala/org/genivi/sota/core/FakeDeviceRegistry.scala [hunk 4]
```diff
@@ -129,4 +129,8 @@ class FakeDeviceRegistry(namespace: Namespace)
   def addGroup(group: Uuid, devices: Seq[Uuid]): Unit = {
     groups.put(group, devices)
   }
+
+  def setInstalledPackages
+  (device: Uuid, json: io.circe.Json)(implicit ec: ExecutionContext) : Future[Unit] =
```
Comment (jerrytrieu) [inline common-client/src/main/scala/org/genivi/sota/core/FakeDeviceRegistry.scala]: Any way to use List instead of json, and still be backward compatible to rvi for now?

File: core/src/main/scala/org/genivi/sota/core/transfer/DeviceUpdates.scala [hunk 5]
```diff
@@ -41,11 +42,15 @@ object DeviceUpdates {
     */
   def update(device: Uuid,
              packageIds: List[PackageId],
-             resolverClient: ExternalResolverClient)
+             resolverClient: ExternalResolverClient,
+             deviceRegistryClient: DeviceRegistry)
             (implicit ec: ExecutionContext): Future[Unit] = {
     // TODO: core should be able to send this instead!
     val j = Json.obj("packages" -> packageIds.asJson, "firmware" -> Json.arr())
```
Comment (jerrytrieu) [inline core/src/main/scala/org/genivi/sota/core/transfer/DeviceUpdates.scala]: Let's try avoiding this kludge for new api

File: common-client/src/main/scala/org/genivi/sota/core/FakeDeviceRegistry.scala (hunk 4 above)
Comment (CalumMcCall) [inline common-client/src/main/scala/org/genivi/sota/core/FakeDeviceRegistry.scala]: The json isn't a list, but a json object with a list of firmware and a list of software. I could move the 'InstalledSoftware' object from resolver into common and use it in core also?

Comment (jerrytrieu) [inline common-client/src/main/scala/org/genivi/sota/core/FakeDeviceRegistry.scala]: Doesn't the device reg api take list of packages? Let's use that.

File: core/src/main/scala/org/genivi/sota/core/transfer/DeviceUpdates.scala (hunk 5 above)
Comment (jerrytrieu) [inline core/src/main/scala/org/genivi/sota/core/transfer/DeviceUpdates.scala]: Pass the packageids to device reg.

File: core/src/main/scala/org/genivi/sota/core/rvi/SotaServices.scala [hunk 6]
```diff
@@ -105,7 +105,10 @@ class SotaServices(updateController: ActorRef, resolverClient: ExternalResolverC
 
   def updatePackagesInResolver( message: InstalledPackages ) : Future[Unit] = {
     log.debug( s"InstalledPackages from rvi: $message" )
-    resolverClient.setInstalledPackages(message.device, message.installed_software)
+    for {
```
Comment (jerrytrieu) [inline core/src/main/scala/org/genivi/sota/core/rvi/SotaServices.scala]: Undo this change. Let resolver pass packages to device reg.

File: device-registry/src/main/scala/org/genivi/sota/device_registry/DevicePackagesResource.scala [hunk 7]
```diff
@@ -49,6 +57,9 @@ class DevicePackagesResource(namespaceExtractor: Directive1[AuthedNamespaceScope
           }
```
Comment (Danten) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/DevicePackagesResource.scala]: The /devices/{:uuid}/packages route seems wrong to me, if we are using `deviceNamespaceAuthorizer` then it will check that the `uuid` is in the extracted namespace. But we are not sending a namespace since this request originate from the device, which doesn't know its namespace. I would prefer to have this in a /mydevice route which doesn't use the deviceNamespaceAuthorizer.

File: core/src/main/scala/org/genivi/sota/core/rvi/SotaServices.scala (hunk 6 above)
Comment (CalumMcCall) [inline core/src/main/scala/org/genivi/sota/core/rvi/SotaServices.scala]: Should I also change the non-rvi invocation to also rely on resolver to pass packages to device reg?

File: web-server/app/org/genivi/webserver/controllers/Application.scala [hunk 8]
```diff
@@ -45,6 +45,7 @@ class Application @Inject() (ws: WSClient,
    */
   def apiByPath(path: String) : String = path.split("/").toList match {
     case "resolver" :: "packages" :: _ :: _ :: "filter" :: _ => resolverApiUri
+    case "package" :: _ :: _ :: "stats" :: _ => deviceRegistryApiUri
```
Comment (Danten) [inline web-server/app/org/genivi/webserver/controllers/Application.scala]: This is not needed anymore right? Because it is in the device route

File: device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala [hunk 9]
```diff
@@ -110,7 +128,13 @@ class DevicesResource(namespaceExtractor: Directive1[AuthedNamespaceScope],
         } ~
         (scope.get & path("groups") & pathEnd) {
           getGroupsForDevice(uuid)
+        } ~
+        (path("packages") & scope.get) {
+          listPackagesOnDevice(uuid)
         }
+      } ~
+      (scope.get & extractPackageId & path("device_count")) { pkg =>
```
Comment (jerrytrieu) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala]: Let's have /api/v1/devices_count instead of subresource of /devices.

File: docs/swagger/sota-device_registry.yml [hunk 10]
```diff
@@ -449,3 +469,12 @@ definitions:
         type: string
         format: dateTime
         description: The last time a package installation was changed
+  PackageStats:
```
Comment (jerrytrieu) [inline docs/swagger/sota-device_registry.yml]: Rename to DevicesCount

File: core/src/main/scala/org/genivi/sota/core/rvi/SotaServices.scala [hunk 11]
```diff
@@ -105,7 +111,14 @@ class SotaServices(updateController: ActorRef, resolverClient: ExternalResolverC
 
   def updatePackagesInResolver( message: InstalledPackages ) : Future[Unit] = {
     log.debug( s"InstalledPackages from rvi: $message" )
-    resolverClient.setInstalledPackages(message.device, message.installed_software)
+    decode[InstalledSoftware](message.installed_software.toString()) match {
```
Comment (jerrytrieu) [inline core/src/main/scala/org/genivi/sota/core/rvi/SotaServices.scala]: Let's not change this file and don't move the firmware type in this PR.

Comment (jerrytrieu) [inline core/src/main/scala/org/genivi/sota/core/rvi/SotaServices.scala]: We'll deal with this in another refactoring of resolver

File: core/src/main/scala/org/genivi/sota/core/rvi/SotaServices.scala [hunk 12]
```diff
@@ -4,15 +4,15 @@
  */
 package org.genivi.sota.core.rvi
 
+import java.util.UUID
```
Comment (Danten) [inline core/src/main/scala/org/genivi/sota/core/rvi/SotaServices.scala]: Why did you move this import?

File: docs/swagger/sota-device_registry.yml [hunk 13]
```diff
@@ -449,3 +469,12 @@ definitions:
         type: string
         format: dateTime
         description: The last time a package installation was changed
+  DevicesCount:
+    type: object
+    properties:
+      deviceCount:
+        type: integer
+        description: count of devices with this package installed
+      groupCount:
```
Comment (Danten) [inline docs/swagger/sota-device_registry.yml]: This is `Set[Uuid]` and not a `integer`.

File: external-resolver/src/test/scala/org/genivi/sota/resolver/test/FirmwareResourceSpec.scala [hunk 14]
```diff
@@ -9,7 +9,7 @@ import java.time.Instant
 import eu.timepit.refined.api.Refined
 import org.genivi.sota.core.FakeDeviceRegistry
 import org.genivi.sota.data.Device.DeviceName
-import org.genivi.sota.data.{Device, Namespace, Namespaces, Uuid}
+import org.genivi.sota.data._
```
Comment (Danten) [inline external-resolver/src/test/scala/org/genivi/sota/resolver/test/FirmwareResourceSpec.scala]: Was this change really necessary?

File: core/src/main/scala/org/genivi/sota/core/rvi/SotaServices.scala (hunk 12 above)
Comment (CalumMcCall) [inline core/src/main/scala/org/genivi/sota/core/rvi/SotaServices.scala]: It was Intellij's auto sort feature. It sorts according to namespace and alphabet. I guess I should probably stop doing this since it just looks confusing in the PRs
